server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -744,6 +755,35 @@ public DataSegment map(int index, ResultSet r, StatementContext ctx) throws SQLE
 
     // Replace "dataSources" atomically.
```

Comment:
"dataSources" and "overshadowedSegments" should be updated atomically, together. Updating them separately means we could get info about overshadowed segments without actually knowing they're overshadowed (or vice versa). The simplest way to do this is to wrap them into a single object and then swap in _that_ object atomically.

One way to do this is to make `getDataSources()` return `Collection<DataSourceSnapshot>` instead of `Collection<ImmutableDruidDataSource>`, and make `DataSourceSnapshot` something that contains both an `ImmutableDruidDataSource` and an `ImmutableSet<SegmentId> overshadowedSegments`. For each one, build the overshadowedSegments list by running `VersionedIntervalTimeline.forSegments(...)` on the segments for that datasource, and then extracting the overshadowed ones.

Reply:
I don't mind if this is implemented, but isn't it a waste of resources since the merge of `DataSegment` and `SegmentWithOvershadowedStatus` objects will make it all obsolete? See https://github.com/apache/incubator-druid/issues/7571#issuecomment-495738096.

Reply:
@leventov aren't you suggesting the same thing in [#7571 (comment)](https://github.com/apache/incubator-druid/issues/7571#issuecomment-495738096), with your first suggestion and your second suggestion
> but this problem naturally goes away itself when SQLSegmentMetadataManager will store SegmentWithOvershadowedStatus objects in dataSources

I'm not sure I understand it, so did you mean to store `SegmentWithOvershadowedStatus`(set of overshadowed segments) in `DruidDataSource` object?

Reply:
created a `DataSourceSnapshot` class which contains a snapshot of druidDatasources and computes the overshadowed segment from that snapshot, this is invoked from `doPoll()`, so both dataSources and overshadowedSegments are updated atomically in every poll cycle. And the getters like `getDataSources` and `getOvershadowedSegments` use this snapshot class to return the values.